### ISSUE TITLE
Review of merged #235 and #262

### ISSUE THREAD

**Issue by gbaydin** [DiffSharp/DiffSharp#258]
@dsyme this merge #235 is slightly worrying. Let's discuss when we meet next time.

About the `Tensor0` case: I understand that you have some very good reason for this based on how you feel the codebase will evolve. I don't like the change, but I have a feeling it needs to happen. `Tensor0` is not a great name. Let's think about a better one `TensorC` can be better. What about `Tensor_`?

Another thing is the "%A" printing which has been relying on the default DU printer. So that we got it look like `Tensor [1.; 2.; 3.]` I have a feeling this can be addressed through other means. (Later edit: I can see now that there is a new tensor printing code merged here #234, will check.)

**More importantly** I have significant worries about the following two changes. I have a feeling that the mistakes you identified were not mistakes but were some intentional design influenced by Python-like conventions. I need to remember and inspect the behavior and then try to understand your change better.

> Fix a mistake in the slicing shape computation where the 1 flag was being set for slices like t.[3..] resulting in incorrect result shapes when using that notaion to slice the first or last single element.

> Fix a mistake in boundsToShape, which is used in two slightly different ways - one to get the result shape of slicing, the other to push adjoints in the SliceTT reverse mode op back into the adjoint for the originating tensor.

Later edit: I think this is also connected with why you added the `check` here
```fsharp
member t.reversePush(value:Tensor) =
    let check (v:Tensor,t:Tensor) = 
        // check the shapes of the adjoints match the nodes to which they are being propagated
        assert (Shape.canExpand v.shape t.derivative.shape || Shape.canExpand t.derivative.shape v.shape)
```

I remember there were some very clear reasons and a careful design for the behavior you changed and it was also about the reverse mode shapes. I'm quite concerned about this merged code and I think the `expand` check is not a good idea. We don't want things in reverse mode to rely on broadcasting/expansions etc. We need to check if the reverse pushed value has exactly the same shape with the primal of the node to which we are propagating. Anything other than that can be a cause for silent erroneous behavior with derivatives in future code. I think it's very likely I would like to revert this merge.

**Comment by dsyme** [comment]
@gbaydin We should talk about this.

> We don't want things in reverse mode to rely on broadcasting/expansions etc. 

Right, I 100% agree, but the code (before and after #235) does rely on broadcasting/expansions I'm afraid. In particular there are places it uses "dsharp.zero()" for the adjoint or derivatives.  

I really think it's much better to assert something about these shapes rather than nothing. It would be better to adjust the assert and make the code tighter, rather than reverting to it being unasserted and implicit as it is now. I'd recommend changing the assert to "is a zero scalar or has the right size"?

Here are some examples where zero values come in:

```
        ...ref (a.zeroLike())...

        let mutable aderivative = a.zeroLike()
        let mutable bderivative = b.zeroLike()
      
        push (check(a.zeroLike(), a) :: tt)
```
In each of these cases the zero scalar relies on expansion. I'm not aware of any other reliance on expansion but there may be others.

> Later edit: I think this is also connected with why you added the check here

IIRC it is connected, but in a positive way that you want - it creates a tensor of the right size, rather than a scalar that is re-broadcast.  I can double check.

Overall the changes here haven't loosened anything - they are correcting things in the way that you want .  In particlar #237 (if we ever accept it) actually needs the shapes to be much tighter in the way that you propose, see [this line](https://github.com/DiffSharp/DiffSharp/pull/237/files#diff-8514611344d9a1f2f2c3bf0e8386bc10d7e347f5225bc69669a7a137d738faccR2881).   It was in the process of tightening the shapes that I found this quirk in the existing code.

**Comment by dsyme** [comment]
Regarding the slicing, here is the inconsistency:

Before 235:
```
> dsharp.ones(4).[2..].shape
val it : Shape = [|2|]

> dsharp.ones(4).[3..].shape
val it : Shape = [||]
```

After 235:
```
> dsharp.ones(4).[2..].shape
val it : Shape = [|2|]

> dsharp.ones(4).[3..].shape
val it : Shape = [|1|]
```

This matches python - the change is correct I'm sure
```
>>> torch.ones(4)[2:].shape;
torch.Size([2])
>>> torch.ones(4)[3:].shape;
torch.Size([1])
```

**Comment by dsyme** [comment]
> About the Tensor0 case: I understand that you have some very good reason for this based on how you feel the codebase will evolve. I don't like the change, but I have a feeling it needs to happen. Tensor0 is not a great name. Let's think about a better one TensorC can be better. What about Tensor_?

I don't mind the name, `TensorC` or `Tensor_` are both fine by me

**Comment by dsyme** [comment]
> Fix a mistake in boundsToShape, which is used in two slightly different ways - one to get the result shape of slicing, the other to push adjoints in the SliceTT reverse mode op back into the adjoint for the originating tensor.

On this: `boundsToShapeNoSqueeze` would be better called `boundsToShapeKeemDims` or simply take a `keepDims` parameter for `boundsToShape`

The need `keepDims=false` is clear for in the call to `boundsToShape` in slicing, that's the point

The need to `keepDims=true` in the reverse propagation for SliceT is pretty logical, here is the call https://github.com/DiffSharp/DiffSharp/pull/235/files#diff-8514611344d9a1f2f2c3bf0e8386bc10d7e347f5225bc69669a7a137d738faccR2936.  Without it the thing being added doesn't have the right number of dimensions to be added into `a.derivative`.  This manifested itself as an `addSlice` shape mismatch when working on #237

**Comment by dsyme** [comment]
@gbaydin https://github.com/DiffSharp/DiffSharp/pull/262 may ease your ill-feeling - it shows that, with the changes in #235, we can tighten the assert to 

            assert (v.shape.Length = 0 || t.derivative.shape.Length = 0 || v.shape = t.derivative.shape)

I think that makes the assert is accurate to the code now (given the use of `zeroLike()` in numerous places, which we can deal with separately)

**Comment by dsyme** [comment]
> Fix a mistake in boundsToShape, which is used in two slightly different ways - one to get the result shape of slicing, the other to push adjoints in the SliceTT reverse mode op back into the adjoint for the originating tensor.

I looked again at this, and there was actually no effective change here.

* boundsToShape was renamed `boundsToShapeNoSqueeze` (it would be better called boundsToShapeKeemDims or simply take a keepDims=true parameter) and the caller also adjusted

* A new, uncalled boundsToShape was introduced corresponding to keepDims=false

So there was actually no change here - this was something I copied over from #237 by mistake.

I have added the following test case to #262 to check gradients of slicing in two dimensions:

```fsharp
        for combo in Combos.AllDevicesAndBackends do

            let revx = combo.tensor([[54.7919; 70.6440; 16.0868; 74.5486; 82.9318]; 
                                     [54.7919; 70.6440; 16.0868; 74.5486; 82.9318]]).reverseDiff()
            let revz = revx.[*,2..2]
            let revzCorrect = combo.tensor([16.0868; 16.0868])
            revz.reverse(combo.tensor([0.9360;0.9360]))
            let revxd = revx.derivative
            let revxdCorrect = combo.tensor([[0.; 0.; 0.9360; 0.; 0.];[0.; 0.; 0.9360; 0.; 0.]])

            Assert.True(revz.allclose(revzCorrect, 0.01))
            Assert.True(revxd.allclose(revxdCorrect, 0.01))
```

**Comment by dsyme** [comment]
BTW here's an example of a test that failed prior to #235 (reverse mode on slice on the final column, fails because of the mismatched shape on the slice)

I've checked this failed prior to #235 and added this test to #262 as well 

```
    [<Test>]
    member _.TestDerivativeSliceT () =
        for combo in Combos.AllDevicesAndBackends do

            let revx = combo.tensor([54.7919; 70.6440; 16.0868]).reverseDiff()
            let revz = revx.[2..]
            let revzCorrect = combo.tensor([16.0868])
            revz.reverse(combo.tensor([0.9360]))
            let revxd = revx.derivative
            let revxdCorrect = combo.tensor([0.; 0.; 0.9360])

            Assert.True(revz.allclose(revzCorrect, 0.01))
            Assert.True(revxd.allclose(revxdCorrect, 0.01))
```

**Comment by dsyme** [comment]
@gbaydin I think #262 should address all your concerns, we can discuss

**Comment by gbaydin** [comment]
> @gbaydin #262 may ease your ill-feeling - it shows that we can tighten the assert to te following without changing any implementation code:
> 
> ```
>         assert (v.shape.Length = 0 || t.derivative.shape.Length = 0 || v.shape = t.derivative.shape)
> ```
> 
> I think that makes the assert is accurate to the code now (given the use of `zeroLike()` in numerous places, which we can deal with separately). There is no other broadcasting going on

Great, I was actually thinking about having the assert in exactly this form (let's call this "good assert"). I agree that having the assert in there is good if it's written that way. So the rule is that we want the backpropagated derivative to be of the exact same shape with the primal value of the node we're propagating to. I would like to bake this in in a very strong way because if this rule doesn't hold, there might be a chance the person implementing some reverse mode op didn't do their homework very well.

My biggest concern was that the other assert (the "bad assert") could be the first step in enabling a silent class of bugs in future ops added to reverse mode. Actually the good assert is actually going to prevent those future bugs I'm worried about. The main problem with these bugs is that they can be completely silent and very serious (machine learning doesn't work, or works sub-par) with no way of debugging what's going on.

**Comment by gbaydin** [comment]
About the location of the assert: if there is no problem I'm missing, can we please have the assert added just once in this line https://github.com/DiffSharp/DiffSharp/blob/830c90961e575c23c105d505988df5460ffdd84e/src/DiffSharp.Core/Tensor.fs#L2684 instead of having these extra `check(...)`s everywhere?

**Comment by dsyme** [comment]
The problem with that location is it doesn't give causality when tracking down a shape mismatch

IIRC I added the assert when tracking down the shape mismatch related to the slicing thing. When it's just before the addition (which will often fail on bad shape anyway) it doesn't help much - you don't know which operator implementation caused the bad shape.

Hence adding it at the point the pairs are pushed

**Comment by dsyme** [comment]
An alternative might be to factor out the mutable worklist to a separate Worklist type and add the asert in the push method for that type. That might have other uses later anyway.

But I do recommend asserting as the tangent/node pairs are produced

**Comment by gbaydin** [comment]
> The problem with that location is it doesn't give causality when tracking down a shape mismatch
I was actually thinking this is the reason behind the choice. Sorry that I forgot to communicate that. 

I had to debug reverse mode implementations when things failed many times. It was difficult to catch where a shape mismatch originates and the only way was to temporarily alter the reverse mode code of the child node (within reversepush) to print extra stuff, inspect the whole graph with [this](https://github.com/DiffSharp/DiffSharp/blob/e832f364cbc6f1f04983db2ce230e150d0887f9f/src/DiffSharp.Core/Tensor.fs#L335) etc. I was going to say "this will be seen only by people who contribute new reverse mode code, which should be very infrequent, and they can also undertake the difficulty of debugging" but I can immediately imagine your answer would be that we should make that experience as strong as we can for long term maintainability. TLDR: I think you're right and this makes the code less prone to problems with future code contributions.

**Comment by dsyme** [comment]
Yup cool. And yes, we should make the experience tight. This is also relevant to #89 (what happens when Reverse returns incorrect sizes)

**Comment by dsyme** [comment]
BTW I've adjusted #89 to be more directly in line with how we specify new operators today. I used it in #252 and #260 and it feels pretty much OK, perhaps initially for internal use or a DiffSharp.Extensions library where people can contribute more freely. 

Feature parity with LibTorch would require 100+ extensions using the fast C++ primal and reverse gradient code. Not all are perf critical but many are. So we need some kind of extensions story

**Comment by dsyme** [comment]
I'll lose this as I think the review feedback is covered in #262

**Comment by gbaydin** [comment]
@dsyme I'm still looking at things and we need to be extremely careful with reaching the correct decision about the assert and the other things involved in this #235 (and the #262 you merged while I was checking things).

I'm not sure whether this is the best assert we can write: `assert (v.shape.Length = 0 || t.derivative.shape.Length = 0 || v.shape = t.derivative.shape)` and I would appreciate if you can give me time and leave the merging to me. I think the rule should be "the backpropagated derivative to be either an empty tensor or of the exact same shape with the primal value of the node we're propagating to" as wrote above, but I need to check this as well.

I also could not follow what's the deal with the `boundsToShape` merge and then the subsequent merge. I need time to check. Can we please stop the merging activity for any code that has to do with the library core logic and derivative handling?

**Comment by gbaydin** [comment]
I'm going to keep my list here as I'm worried I can forget things:
- Reverse mode assert in #262: the merged assert needs to be improved
- Slicing changes merged in #235: I'm still inspecting but I think you're right about the merged fix is correcting a mistake
- Bounds-related changes in #235 and #262: still no idea but I think I will understand when I read your messages more carefully
- Tests merged in #262: I need to check the derivative tests that got merged and we need to add forward derivative tests
- Tests merged in #262: We need to add at least one test for the >1 dimensional case

**Comment by dsyme** [comment]
OK no problem 

> Bounds-related changes in #235 and #262: still no idea but I think I will understand when I read your messages more carefully

This was basically reverted in 262. There's no change here now end to end

**Comment by dsyme** [comment]
> Reverse mode assert in #262 <https://github.com/DiffSharp/DiffSharp/pull/262>: the merged assert needs to be improved

Yes I think it would be good to be stronger. The zero values don't seem to serve much role and maybe just make them zerosLike according to the required shape. 


Tbh when working on #237 (which is non trivial) I came to the conclusion that the assert should  be shape equality (especially if the derivatives are going to be mutable tensor registers).

**Comment by gbaydin** [comment]
@dsyme I started #263 to address the issues we've been talking about. Let's put our changes there and merge when we both give it the green light.

**Comment by gbaydin** [comment]
#263 is merged. Closing the issue.
